src/dmd/backend/barray.d

```diff
@@ -31,27 +31,29 @@ struct Barray(T)
      */
     void setLength(size_t length)
     {
-        if (capacity < length)
+        static void enlarge(ref Barray barray, size_t length)
         {
-            if (capacity == 0)
-                capacity = length;
-            else
-                capacity = length + (length >> 1);
-            capacity = (capacity + 15) & ~15;
-            T* p = cast(T*)realloc(array.ptr, capacity * T.sizeof);
+            pragma(inline, false);
+            auto newcap = (barray.capacity == 0) ? length : length + (length >> 1);
+            barray.capacity = (newcap + 15) & ~15;
+            T* p = cast(T*)realloc(barray.array.ptr, barray.capacity * T.sizeof);
             if (length && !p)
             {
                 version (unittest)
                     assert(0);
                 else
                     err_nomem();
             }
-            array = p[0 .. length];
+            barray.array = p[0 .. length];
         }
+
+        if (length < capacity)
```

Comment:
`<=` rather than `<` or the slow path will be taken unnecessarily when the array is shrunk to zero then expanded back to the same size.

Reply:
fixed